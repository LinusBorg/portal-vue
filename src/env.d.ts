/// <reference types="vite/client" />

declare const __PORTAL_VUE_VERSION__: string
declare const __NODE_ENV__: string
