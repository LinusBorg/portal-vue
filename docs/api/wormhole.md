---
sidebar: auto
prev: ./mounting-portal
next: false
---

# Wormhole Object

The Wormhole is not a component, it's an object that connects the `<Portal>`s to their `<PortalTarget>`s.

Usually, you will never need to use this object, but you _can_ use this object to programmatically send content to a `<PortalTarget>`, check whether a target exists, and other stuff.

:::warning Public API
The wormhole object exposes quite a few properties and methods. With regard to semver, only the properties and methods documented below are considered part of the public API.

Use any other properties at your own risk.
:::

## Methods

### open()

The `open` method accepts one argument, an object with the following properties:

`Wormhole.open({to, from, passengers})`

| Property   | Required | Default  | Explanation                                                         |
| ---------- | -------- | -------- | ------------------------------------------------------------------- |
| to         | yes      |          | The name of the `<PortalTarget>` to send to                         |
| from       | yes      |          | The name of the `<Portal>` this content comes from.                 |
| passengers | yes      |          | An array of vNodes - the content to be sent to the `<PortalTarget>` |
| order      | no       | Infinity | a number indicating the order when multipe sources for a target are used |

Even if you use this method programmatically and there is not source `<Portal>`, you still have to provide `from` - every content sent through the wormhole needs a source.

Examples:

```javascript
import { Wormhole } from 'portal-vue'

methods: {
  sendToPortalTarget() {
    const passengers = [this.$createElement('p', 'This will be displayed in the Target!')]
    Wormhole.open({
      to: 'destination',
      from: 'my-portal',
      passengers,
    })
  }
}
```

This is the programmatic equivalent of the following:

<!-- prettier-ignore -->
```html
<portal to="destination" name="my-portal">
  <p>This will be displayed in the Target!</p>
</portal>
```

::: tip Server-Side Rendering
For the reasons layed out in [the section about SSR](../guide/SSR.md), this method won't do anything during Server-Side Rendering. Portal'ing of the content will only happen on the client.
Make sure to Read the linked section for more information.
:::

### close()

As the name suggests, this is the counterpart to `open()`. It's used to remove content from a `<PortalTarget>`.

It accepts two arguments, an object and an optional `force` flag.

`Wormhole.close({ to, from }, force = false)`

| Argument/Property | Required | Default | Explanation                                                                                      |
| ----------------- | -------- | ------- | ------------------------------------------------------------------------------------------------ |
| to                | yes      |         | The name of the `<PortalTarget>` whose content should be removed                                 |
| from              | no       |         | Name of the Portal that the content was created with. Can be left out if the `force` flag is set |
| force             | no       | false   | If this is true, the `<PortalTarget>`s content will be removed even if `from` doesn't match      |

Why do you have to provide the `from` name (or use `force`)?

The close() method is also used by `<Portal>` components to remove the content from a `<PortalTarget>` when the `<Portal>` is being destroyed. But that should only remove the content if it still actually is the content from that `<Portal>`. If another `<Portal>` have overridden our content before the first one gets destroyed, we have to keep the content instead of removing it.

We can override this behaviour with the `force` argument.

Example:

```javascript
methods: {
  clearPortalTarget() {
    Wormhole.close({
      to: 'destination',
      from: 'my-programmatic-portal'
    })

    // Variation with `force`
    Wormhole.close({
      to: 'destination',
    }, true)

    // if the `from` doesn't match the source the content
    // was sent from, close() will fail silently
    // IF the current content is from 'my-other-portal', then:
    Wormhole.close({
      to: 'destination',
      from: 'my-portal',
    })
    // => this will not remove any content, it will stay in the PortalTarget
  }
}
```

### hasSource() <Badge text="2.0.0+"/>

`Wormhole.hasSource(from)`

Returns `true` if there's a `<Portal>` with the given name active.

Example:

```javascript
Wormhole.hasSource('origin')
// => true/false
```

::: tip Server-Side Rendering
For the reasons layed out in [the section about SSR](../guide/SSR.md), this method will always return `false` during Server-Side Rendering.
Make sure to Read the linked section for more information.
:::

### hasTarget() <Badge text="changed in 2.0.0" type=warning />

`Wormhole.hasTarget(to)`

Returns `true` if there's a `<PortalTarget>` with the given name active.

Example:

```javascript
Wormhole.hasTarget('destination')
// => true/false
```

::: tip Server-Side Rendering
For the reasons layed out in [the section about SSR](../guide/SSR.md), this method will always return `false` during Server-Side Rendering.
Make sure to Read the linked section for more information.
:::

### hasContentFor() <Badge text="changed in 2.1.0" type=warning />

`Wormhole.hasContentFor(to)`

Returns `true` if there's a `<PortalTarget>` with a name of `to` _and_ it currently has content to show.

Example:

```javascript
Wormhole.hasContentFor('destination')
// => true/false
```

::: tip Server-Side Rendering
For the reasons layed out in [the section about SSR](../guide/SSR.md), this method will always return `false` during Server-Side Rendering.
Make sure to Read the linked section for more information.
:::
