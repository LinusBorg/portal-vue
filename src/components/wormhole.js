import Vue from 'vue'
import { combinePassengers, freeze, stableSort } from '../utils'
const transports = {}

export { transports }

export const Wormhole = Vue.extend({
  data: () => ({ transports }),
  methods: {
    open(transport) {
      const { to, from, passengers } = transport
      if (!to || !from || !passengers) return

      transport.passengers = freeze(passengers)
      const keys = Object.keys(this.transports)
      if (keys.indexOf(to) === -1) {
        Vue.set(this.transports, to, [])
      }

      const currentIndex = this.getTransportIndex(transport)
      // Copying the array here so that the PortalTarget change event will actually contain two distinct arrays
      const newTransports = this.transports[to].slice(0)
      if (currentIndex === -1) {
        newTransports.push(transport)
      } else {
        newTransports[currentIndex] = transport
      }
      this.transports[to] = stableSort(newTransports, function(a, b) {
        return a.order - b.order
      })
    },

    close(transport, force = false) {
      const { to, from } = transport
      if (!to || !from) return
      if (!this.transports[to]) {
        return
      }

      if (force) {
        this.transports[to] = []
      } else {
        const index = this.getTransportIndex(transport)
        if (index >= 0) {
          // Copying the array here so that the PortalTarget change event will actually contain two distinct arrays
          const newTransports = this.transports[to].slice(0)
          newTransports.splice(index, 1)
          this.transports[to] = newTransports
        }
      }
    },

    hasTarget(to) {
      return this.transports.hasOwnProperty(to)
    },

    hasContentFor(to) {
      if (!this.transports[to]) {
        return false
      }
      return this.getContentFor(to).length > 0
    },

    getSourceFor(to) {
      return this.transports[to] && this.transports[to][0].from
    },

    getContentFor(to) {
      const transports = this.transports[to]
      if (!transports) {
        return undefined
      }
      return combinePassengers(transports)
    },

    getTransportIndex({ to, from }) {
      for (const i in this.transports[to]) {
        if (this.transports[to][i].from === from) {
          return i
        }
      }
      return -1
    },
  },
})

const wormhole = new Wormhole(transports)
export default wormhole
