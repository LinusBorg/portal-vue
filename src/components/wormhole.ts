import Vue from 'vue'
import { freeze, inBrowser, stableSort } from '../utils'
import {
  Transports,
  Transport,
  TransportInput,
  TransportVector,
  VMRegister,
} from '../types'

const transports: Transports = {}
const targets: VMRegister = {}
const sources: VMRegister = {}

export const Wormhole = Vue.extend({
  data: () => ({
    transports,
    targets,
    sources,
    trackInstances: inBrowser,
  }),
  methods: {
    open(transport: TransportInput) {
      if (!inBrowser) return
      const { to, from, passengers, order = Infinity } = transport
      if (!to || !from || !passengers) return

      const newTransport = {
        to,
        from,
        passengers: freeze<object>(passengers),
        order,
      } as Transport
      const keys = Object.keys(this.transports)

      if (keys.indexOf(to) === -1) {
        Vue.set(this.transports, to, [])
      }

      const currentIndex = this.$_getTransportIndex(newTransport)
      // Copying the array here so that the PortalTarget change event will actually contain two distinct arrays
      const newTransports = this.transports[to].slice(0)
      if (currentIndex === -1) {
        newTransports.push(newTransport)
      } else {
        newTransports[currentIndex] = newTransport
      }

      this.transports[to] = stableSort<Transport>(
        newTransports,
        (a: Transport, b: Transport) => a.order - b.order
      )
    },

    close(transport: TransportVector, force = false) {
      const { to, from } = transport
      if (!to || (!from && force === false)) return
      if (!this.transports[to]) {
        return
      }

      if (force) {
        this.transports[to] = []
      } else {
        const index = this.$_getTransportIndex(transport)
        if (index >= 0) {
          // Copying the array here so that the PortalTarget change event will actually contain two distinct arrays
          const newTransports = this.transports[to].slice(0)
          newTransports.splice(index, 1)
          this.transports[to] = newTransports
        }
      }
    },
    registerTarget(target: string, vm: Vue, force?: boolean): void {
      if (!inBrowser) return
      if (this.trackInstances && !force && this.targets[target]) {
        console.warn(`[portal-vue]: Target ${target} already exists`)
      }
      this.$set(this.targets, target, Object.freeze([vm]))
    },
    unregisterTarget(target: string) {
      this.$delete(this.targets, target)
    },
    registerSource(source: string, vm: Vue, force?: boolean): void {
      if (!inBrowser) return
      if (this.trackInstances && !force && this.sources[source]) {
        console.warn(`[portal-vue]: source ${source} already exists`)
      }
      this.$set(this.sources, source, Object.freeze([vm]))
    },
    unregisterSource(source: string) {
      this.$delete(this.sources, source)
    },
    hasTarget(to: string) {
      return !!(this.targets[to] && this.targets[to][0])
    },
    hasSource(to: string) {
      return !!(this.sources[to] && this.sources[to][0])
    },
    hasContentFor(to: string) {
      return !!this.transports[to] && !!this.transports[to].length
    },
    // Internal
    $_getTransportIndex({ to, from }: TransportVector): number {
      for (const i in this.transports[to]) {
        if (this.transports[to][i].from === from) {
          return +i
        }
      }
      return -1
    },
  },
})

const wormhole = new Wormhole(transports)
export { wormhole }
