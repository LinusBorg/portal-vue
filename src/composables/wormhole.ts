import { type InjectionKey, inject, provide } from 'vue'
import type { Wormhole } from '../types'

export const wormholeSymbol = Symbol('wormhole') as InjectionKey<Wormhole>

export function useWormhole() {
  const wh = inject(wormholeSymbol)

  if (!wh) {
    throw new Error(`
    [portal-vue]: Necessary Injection not found. Make sur you installed the plugin properly.`)
  }

  return wh
}

export function provideWormhole(wormhole: Wormhole) {
  provide(wormholeSymbol, wormhole)
}
