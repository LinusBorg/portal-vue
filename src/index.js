import Portal from './components/portal.js'
import PortalTarget from './components/portal-target.js'
import Wormhole from './components/wormhole.js'

function install(Vue, opts = {}) {
  Vue.component(opts.portalName || 'Portal', Portal)
  Vue.component(opts.portalTargetName || 'PortalTarget', PortalTarget)
}
if (typeof window !== 'undefined' && window.Vue) {
  window.Vue.use({ install: install })
}

export default {
  install,
  Portal,
  PortalTarget,
  Wormhole,
}
