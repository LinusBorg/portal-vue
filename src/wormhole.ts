import { reactive, readonly } from 'vue'
import type {
  Name,
  Transport,
  TransportCloser,
  TransportInput,
  TransportsHub,
  Wormhole,
} from './types'
import { inBrowser, stableSort } from './utils'

export function createWormhole(asReadonly = true): Wormhole {
  const transports: TransportsHub = reactive(new Map())

  function open(transport: TransportInput) {
    if (!inBrowser) return

    const { to, from, content, order = Infinity } = transport
    if (!to || !from || !content) return

    if (!transports.has(to)) {
      transports.set(to, new Map())
    }
    const transportsForTarget = transports.get(to)!

    const newTransport = {
      to,
      from,
      content,
      order,
    } as Transport

    transportsForTarget.set(from, newTransport)
  }

  function close(transport: TransportCloser) {
    const { to, from } = transport
    if (!to || !from) return
    const transportsForTarget = transports.get(to)
    if (!transportsForTarget) {
      return
    }
    transportsForTarget.delete(from)
    if (!transportsForTarget.size) {
      transports.delete(to)
    }
  }

  function getContentForTarget(target: Name, returnAll?: boolean) {
    const transportsForTarget = transports.get(target)
    if (!transportsForTarget) return []

    const content = Array.from(transportsForTarget?.values() || [])

    if (!returnAll) {
      // return Transport that was added last
      return [content.pop()] as Transport[]
    }
    // return all Transports, sorted by their order property
    return stableSort(
      content,
      (a: Transport, b: Transport) => a.order - b.order
    )
  }

  const wh: Wormhole = {
    open,
    close,
    transports,
    getContentForTarget,
  }
  return asReadonly ? (readonly(wh) as Wormhole) : wh
}

export const wormhole = createWormhole()
