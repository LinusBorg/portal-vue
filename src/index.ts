import { App } from 'vue'
import Portal from './components/portal'
import PortalTarget from './components/portal-target'
import {
  provideWormhole,
  useWormhole,
  wormholeSymbol,
} from './composables/wormhole'
import { Wormhole as TWormhole } from './types'
import { createWormhole, wormhole as defaultWormhole } from './wormhole'
export { mountPortalTarget } from './utils/mountPortalTarget'
export interface PluginOptions {
  portalName?: string | false
  portalTargetName?: string | false
  MountingPortalName?: string
  wormhole?: TWormhole
}

export function install(app: App, options: PluginOptions = {}) {
  options.portalName !== false &&
    app.component(options.portalName || 'Portal', Portal)
  options.portalTargetName !== false &&
    app.component(options.portalTargetName || 'PortalTarget', PortalTarget)

  const wormhole = options.wormhole ?? defaultWormhole
  app.provide(wormholeSymbol, wormhole)
}

// alternative name for named import
export const plugin = install
export const Wormhole = defaultWormhole

export {
  Portal,
  PortalTarget,
  useWormhole,
  provideWormhole,
  TWormhole,
  createWormhole,
}
