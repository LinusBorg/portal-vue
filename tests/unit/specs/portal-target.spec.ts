import { Slot, h, nextTick } from 'vue'
import { mount } from '@vue/test-utils'
import PortalTarget from '../../../src/components/portal-target'
import { wormholeSymbol } from '../../../src/composables/wormhole'
import { createWormhole } from '../../../src/wormhole'

const createWormholeMock = () => {
  const wh = createWormhole(false)

  const open = jest.spyOn(wh, 'open')
  const close = jest.spyOn(wh, 'close')

  return wh
}
// Utils

function createWrapper(props = {}, options = {}) {
  const wh = createWormholeMock()
  return {
    wh,
    wrapper: mount(PortalTarget, {
      props: {
        name: 'target',
        ...props,
      } as any,
      global: {
        provide: {
          [(wormholeSymbol as unknown) as string]: wh,
        },
      },
      ...options,
    }),
  }
}

function generateSlotFn(text = '') {
  return (() => h('div', { class: 'testnode' }, text) as unknown) as Slot
}

describe('PortalTarget', () => {
  it('renders a single element for single vNode with slim prop & single slot element', async () => {
    const { wrapper, wh } = createWrapper()
    const content = generateSlotFn()
    wh.open({
      from: 'source',
      to: 'target',
      content,
    })

    await nextTick()

    expect(wrapper.html()).toBe(
      '<div style="display: none;"></div><div class="testnode"></div>'
    )
  })

  it('renders slot content when no other content is available', function () {
    const { wrapper } = createWrapper(
      {},
      {
        slots: {
          default: `<p class="default">Test</p>`,
        },
      }
    )
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('p.default').exists()).toBe(true)
  })

  it('emits change event', async () => {
    const { wrapper, wh } = createWrapper()

    await nextTick()

    const content = generateSlotFn()
    wh.open({
      to: 'target',
      from: 'source',
      content,
    })

    await nextTick()

    expect(wrapper.emitted<any>().change[0]).toMatchObject([
      {
        hasContent: true,
        sources: ['source'],
      },
    ])
  })
})
