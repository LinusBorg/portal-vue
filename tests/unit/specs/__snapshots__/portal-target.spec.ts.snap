// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortalTarget renders slot content when no other content is available 1`] = `"<p class=\\"default\\">Test</p>"`;
